### DISCUSSION TITLE
How do I reference “params”?

### DISCUSSION THREAD

**Discussion by hmltn-0** [fastapi/fastapi#6319] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
This is the output of this test:

poetry run pytest -xsk task1

async_client = <httpx.AsyncClient object at 0x1076cb190>

    @pytest.mark.asyncio
    async def test_task1(async_client):
        response = await async_client.get("/task1/greet/Jasmin")
        assert response.status_code == 200
        assert response.json() == "Hallo Jasmin, ich bin Emilia."
    
        response = await async_client.get("/task1/greet/Stefan", params={"language": "en"})
        assert response.status_code == 200
>       assert response.json() == "Hello Stefan, I am Emilia."
E       AssertionError: assert 'Hallo Stefan...h bin Emilia.' == 'Hello Stefan, I am Emilia.'
E         - Hello Stefan, I am Emilia.
E         ?  ^            ^^^^
E         + Hallo Stefan, ich bin Emilia.
E         ?  ^            ^^^^^^^

test_emilia.py:22: AssertionError

I am supposed to edit this code:

@app.get("/task1/greet/{name}", tags=["Task 1"], summary="👋🇩🇪🇬🇧🇪🇸")
async def task1_greet(name: str) -> str:
    """Greet somebody in German, English or Spanish!"""
    
    return f"Hallo {name}, ich bin Emilia."

I don’t understand how to reference the “params={“language”:”en”}” in the test above. Assuming it’s being sent by the method “async_client.get()”,how do I pick that up in the method “task1_greet”?

Do I make a method argument like **kwargs, perhaps?

Thanks very much
```

### Description

I need to write code that returns text depending on the language passed in “params”, and I don’t know how to reference “params” because I don’t understand how async_client.get corresponds to async def task1_greet.

I will explore kwargs but I’d still like to understand better how these two methods interact with each other.

Thank you 

### Operating System

macOS

### FastAPI Version

latest

### Python Version

latest

**Comment by AlrasheedA** [comment, marked as answer]
Hi @peterelbert, params are passed as [query parameters](https://fastapi.tiangolo.com/tutorial/query-params/). 

Example usage in your case: 
```python
from fastapi import FastAPI 

app = FastAPI()

@app.get("/task1/greet/{name}")
async def task1_greet(name: str, language: str = None) -> str:
    """Greet somebody in German, English or Spanish!"""
    if language == "en":
        return f"Hello {name}, I am Emilia."
    else: # Default 
        return f"Hallo {name}, ich bin Emilia."

from fastapi.testclient import TestClient
client = TestClient(app)
client.get('/task1/greet/Jasmin', params={'language': 'en'}).json()
```
Gives:
```python
'Hello Jasmin, I am Emilia.'
```

**Comment by hmltn-0** [comment]
Thank you very much.
I think I tried that but for the first test they don’t pass any params, so “language” would be/was unrecognised.

Is there a way to make the variable optional in some way?

Thank you

**Comment by rogerbrinkmann** [comment]
@peterelbert , take a look at this line of AlrasheedA's code:
```python
async def task1_greet(name: str, language: str = None) -> str:
```
`str` gets the default value of `None`, making it optional, in case the `language` parameter isn't set.

**Comment by hmltn-0** [comment]
Awesome, thank you, that works

**Comment by AlrasheedA** [comment]
@peterelbert Please close the issue if it worked.
